[Chen Jia Jun] iP

iamchenjiajun opened the pull request.

iamchenjiajun committed:
Add greeting on launch (Level-0)

iamchenjiajun committed:
Add REPL to process user input
Echoes all user commands except "bye" which terminates the program.

iamchenjiajun committed:
Update divider text
This is done to customize the chatbot.

iamchenjiajun committed:
Add ability to store and list text

iamchenjiajun committed:
Move logo display string into greet()

Greeting logo string is in the main function, which has similar
functionality as greet().

This groups the greetings into a single function for improved
readability in the main function.

iamchenjiajun committed:
Add the ability to mark tasks as done

Tasks are represented by strings stored in an array.

Functionality and data of tasks are separated and hard to extend.

Abstracting each task into the Task class allows functionality to be
grouped with data. This improves code readability and makes it easier
to add tasks related features.

iamchenjiajun committed:
Remove unused function echo()

iamchenjiajun committed:
Update code to comply with standard

iamchenjiajun committed:
Add toString() method to Task

There are instances where Task needs to be printed with
the same formatting.

By using overriding toString() method, the string's formatting is
refactored to a single function in the codebase.

This reduces code duplication and enhances code readability,
allowing it to be modified from a single function in the code.
It can also be overridden by subclasses for different formatting.

iamchenjiajun committed:
Add support for tracking three types of tasks.

Task is being used to to track user's tasks.

There may be different types of tasks that users want to store,
such as todos, deadlines and events.

Three different types of tasks are added as subclasses to Task.
These classes are Todo, Deadline and Event, which has additional
properties to track additional information about these Tasks.

Polymorphic code allows us to use a single collection to store
all the user's Tasks, and use the call the same methods, thus
simplifying the code.

iamchenjiajun committed:
Add automated text UI testing with I/O redirection

Symbols that indicate status icons for Tasks are emojis, which
caused encoding errors in testing output.

The symbols have been replaced with other characters to remove
the cause of these errors.

iamchenjiajun committed:
Update REPL to use switch statement for commands

The loop uses if-else statements to match user commands. As
more commands are added, the structure starts to become messy.

Using a switch statement to match the first word allows us to label
each part of the code more clearly to improve code readability.

iamchenjiajun committed:
Update print statement for event

Expected test output was also updated to match the
updated print statement.

iamchenjiajun committed:
Move Task related functionality into TaskManager

Both REPL and Task related functionality and data exists in the Duke
class, making the code disorganized.

Task related functionality and data has been abstracted into
TaskManager, which manages all Task functionality, such as adding
Tasks.

By abstracting code into Task Manager, the main function is cleaned
up and related functionality is grouped into a single class.

iamchenjiajun committed:
Refactor max tasks size as a constant

iamchenjiajun committed:
Refactor command strings as constants

shaojingle commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
+
 public class Duke {
-    public static void main(String[] args) {
+    public static final String COMMAND_LIST = "list";
```
I think upper case is used for constants and strings are immutable by default. Thus, maybe can change the naming of the variables to fit coding standard.

shaojingle reviewed (commented):
Generally well done!

bqxy commented on src/main/java/Duke.java:
```diff
@@ -1,6 +1,13 @@
 
 public class Duke {
+    public static final String COMMAND_LIST = "list";
+    public static final String COMMAND_BYE = "bye";
+    public static final String COMMAND_DONE = "done";
+    public static final String COMMAND_ADD_TODO = "todo";
+    public static final String COMMAND_ADD_DEADLINE = "deadline";
+    public static final String COMMAND_ADD_EVENT = "event";
+
```
Good idea for refactoring!

qkbrandon reviewed (commented):
Looks good! 👍

xuche123 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,51 @@
+        for (int i = 0; i < tasksCount; i++) {
+            int index = i + 1;
+            System.out.println(index + "." + tasks[i]);
+        }
+    }
+
+    public void markAsDone(int index) {
+        tasks[index].markAsDone();
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println(tasks[index]);
+    }
+
+    public void addTask(String description) {
+        tasks[tasksCount] = new Task(description);
+        System.out.println("Added " + description + " as a Task.");
+        System.out.println(tasks[tasksCount]);
+        tasksCount++;
+    }
+
+    public void addDeadline(String description, String by) {
+        tasks[tasksCount] = new Deadline(description, by);
+        System.out.println("Added " + description + " as a Deadline.");
+        System.out.println(tasks[tasksCount]);
+        tasksCount++;
+    }
+
+    public void addTodo(String description) {
+        tasks[tasksCount] = new Todo(description);
+        System.out.println("Added " + description + " as a Todo.");
+        System.out.println(tasks[tasksCount]);
+        tasksCount++;
+    }
+
+    public void addEvent(String description, String at) {
+        tasks[tasksCount] = new Event(description, at);
+        System.out.println("Added " + description + " as an Event.");
+        System.out.println(tasks[tasksCount]);
+        tasksCount++;
+    }
+}
```
The creation of this separate class makes the main Duke file much more clean and readable. Good idea!

qkbrandon replied:
Good idea to create this class; improves the code structure and helps with readability in the main class

qkbrandon commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) {
+        this.description = description;
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "[✓]" : "[✗]");
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    @Override
+    public String toString() {
+        return getStatusIcon() + " " + getDescription();
+    }
+}
```
Variables and classes properly described, easy to understand, good job!

qkbrandon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,80 @@
 public class Duke {
-    public static void main(String[] args) {
+    public static final String COMMAND_LIST = "list";
+    public static final String COMMAND_BYE = "bye";
+    public static final String COMMAND_DONE = "done";
+    public static final String COMMAND_ADD_TODO = "todo";
+    public static final String COMMAND_ADD_DEADLINE = "deadline";
+    public static final String COMMAND_ADD_EVENT = "event";
```
Good idea for refactoring; helps with readability

qkbrandon reviewed (commented):
Looks good! Good job!

iamchenjiajun committed:
Refactor Duke's output strings as constants

iamchenjiajun committed:
Add error handling for incorrect inputs

The program has no way of handling invalid input from users, and usually
results in the program crashing. We want Duke to handle expected invalid
user input, so that the program can continue to run. Incorrect inputs by
users are invalid commands, invalid argument length, and invalid
arguments (such as giving 'done' followed by a non-integer, or an out of
range value).

Three exception classes were added, two of which extends DukeException.
InvalidArgumentException is thrown when the user's argument is invalid,
and UnknownCommandException is thrown when the user's command is
invalid. The switch statement for handling commands has been moved to
parseCommand().

This allows us to separate the main working code from the error handling
code which maintains code readability and better follow the SLAP
principles.

A getter function was also added to get the current tasks count in
TaskManager.

iamchenjiajun committed:
Update Task to abstract class

iamchenjiajun committed:
Refactor error string into Duke as constant

This puts all the error strings into one place in the code, making the
strings easier to modify.

iamchenjiajun committed:
Update error strings in Duke

iamchenjiajun committed:
Update tests to test error handling

iamchenjiajun committed:
Merge branch 'branch-Level-5'

iamchenjiajun committed:
Divide classes into packages

All classes are not under any packages, and there is a moderate amount
of files in the same directory.

This makes the code hard to manage as files and the directory structure
does not give any indication to the classes that are being used in the
project.

A duke package is created, which contains two other packages: duke.task
and duke.exceptions, to contain the two classes, Task and DukeException,
as well as their subclasses, that are being used in the main Duke
program.

This allows the classes to be organised into packages for easier
management and the directory structure gives us information about the
classes being used in the project. By grouping related classes together,
the code is easier to navigate, maintain and extend.

iamchenjiajun committed:
Update runtest.bat to match new file structure

iamchenjiajun committed:
Merge branch 'branch-A-Packages'

iamchenjiajun committed:
Update runtest.bat to match file structure

iamchenjiajun committed:
Update TaskManager to use ArrayList to store Tasks

Tasks are stored in a fixed size Array, and only allows the user to
store a limted amount of tasks.

TaskManager is updated to store Tasks with an ArrayList.

By storing Tasks with an ArrayList, the user is able to store an
indefinite amount of Tasks. It also removes the need to store the
size of the Array with tasksCount, as the ArrayList provides a
method to access the current size of the list.

iamchenjiajun committed:
Add support for deleting tasks from the list

TaskManager does not have support for deleting Tasks.

A new method for deleting Tasks was added to TaskManager, which removes
the Task using ArrayList's remove method. A new command was added to
Duke, which calls the new method in TaskManager to remove the Task
at the given index.

This allows the user to remove the task, and TaskManager handles the
Tasks related functionality, while Duke handles the command related
functionality.

iamchenjiajun committed:
Update .gitignore to ignore data files

iamchenjiajun committed:
Add saving and loading tasks from file

FileFormatException is thrown when the user's file format is invalid
Task class has an abstract method toSaveString, which is called
during saving of data to file.
loadTasks, saveTasks and checkFileExists are new methods in TaskManager
that handle the loading, saving and creation of data files.

iamchenjiajun committed:
Update expected test results

iamchenjiajun committed:
Update test expected input and output

iamchenjiajun committed:
Merge branch 'branch-Level-6'

iamchenjiajun committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/TaskManager.java

Retain most changes in Duke.java
Retain most changes in TaskManager.java

iamchenjiajun committed:
Update .gitignore to ignore data files in test

manaswini05 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,177 @@
+        System.out.println(DUKE_BYE);
+    }
+
+    public static void checkArgumentsLength(int argumentLength, int expectedLength, String errorMessage)
+            throws InvalidArgumentException {
+        if (argumentLength < expectedLength) {
+            throw new InvalidArgumentException(errorMessage);
```
Be careful with indentation here

iamchenjiajun replied:
According to the Java coding standards indicated in the course materials, shouldn't the wrapped line have 8 more spaces than the parent line?

"Indentation for wrapped lines should be 8 spaces (i.e. twice the normal indentation of 4 spaces) more than the parent line."

https://se-education.org/guides/conventions/java/basic.html#layout

manaswini05 replied:
Yes my bad. Overlooked that it was a wrapped line in a hurry!

manaswini05 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,177 @@
+        }
+    }
+
+    public static void checkValidIntegerRange(int checkInteger, int expectedValue, String errorMessage)
+            throws InvalidArgumentException {
```
Check your indentation

manaswini05 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,177 @@
+        }
+    }
+
+    public static void parseCommand(String line) throws UnknownCommandException, InvalidArgumentException {
```
The method seems too long. Consider extracting some of your processing in the switch-case commands into new methods

manaswini05 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,177 @@
+            System.out.println(ERROR_TASK_FORMAT);
+        }
+
+        // parse and execute commands
```
Put the comments in a more formal way (in other lines as well)

manaswini05 reviewed (commented):
Overall, great work! :) Do look into your commenting style

iamchenjiajun committed:
Refactor Duke to run on an instance in main

iamchenjiajun committed:
Update methods in Duke to non-static

Currently, methods in Duke are static, and static methods should be
avoided in an OOP application as they are considered to be procedural.

The static keyword has been removed from all methods in the Duke class.

As code has been previously refactored into the run method, and Duke
runs in an instance, we no longer need static methods and static
variables, which makes the code more testable.

iamchenjiajun committed:
Refactor Ui class to deal with user interaction

iamchenjiajun committed:
Update build directories in runtest.bat

iamchenjiajun committed:
Refactor Storage class to load and save tasks

iamchenjiajun committed:
Update build directories in runtest.bat

This is done to include the new package that contains the Storage class

iamchenjiajun committed:
Refactor error strings and read command into Ui

iamchenjiajun committed:
Refactor Parser class to parse user input

iamchenjiajun committed:
Update build directories in runtest.bat

This includes the package containing the new Parser class.

iamchenjiajun committed:
Refactor commands into Command classes

iamchenjiajun committed:
Refactor TaskManager into its own package

This makes the entry point of the program more obvious as there is now
only 1 file in the highest package directory. The imports in other
source files are also updated to the correct package.

iamchenjiajun committed:
Update build directories in runtest.bat

This includes the package containing the new Command class and its
subclasses, as well as the new package containing TaskManager.

iamchenjiajun committed:
Update .gitignore to ignore MANIFEST.MF

iamchenjiajun committed:
Refactor creating commands into separate functions

The switch statement in Parser.parseCommand is too long, and thus the
readability of the code is reduced. This also causes the function to
be lengthy.

Let's refactor non-trivial functionality of the switch statement into
its own function that returns a Command.

This makes the code more readable and adhere to SLAP principles.

Checking of validity of indexes in arguments such as done and delete is
refactored into TaskManager from Parser as it is a functionality that
should be managed by the TaskManager instead.

iamchenjiajun committed:
Remove redundant parameters in TaskManager methods

Some functions in TaskManager methods have a boolean parameter
shouldPrintMessage that decides if the output should be printed to the
user. This was previously added as loading from file uses these methods
and thus the output should not be shown to the user. However, this is
now deprecated and should be removed from the code.

This boolean parameter has been removed from methods in TaskManager that
contain it.

The removal of deprecated code leads to better readability.

iamchenjiajun committed:
Add Parser method to check exact argument length

checkArgumentsLength has been renamed to checkMinimumArgumentsLength to
make the method name more descriptive.

iamchenjiajun committed:
Add new strings for errors when deleting tasks

iamchenjiajun committed:
Update expected test results for new strings

iamchenjiajun committed:
Add exception to handle invalid indexes

iamchenjiajun committed:
Refactor expected/min argument length as constants

Expected and minimum argument length for each command are used in the
code as magic numbers, which may confuse readers of the code.

These have been refactored as constants in their corresponding Command
subclass.

This improves readability of the code by removing magic numbers.

checkMinimumArgumentsLength and checkExactArgumentsLength has been
renamed to checkMinArgumentLength and checkExactArgumentLength to
better reflect checking a singular argument string, and to make the
variable name shorter.

iamchenjiajun committed:
Update Deadline and Event to accept date and time

Deadline and Event stores the deadline and event as strings, which is
fine for the current set of functionality but may be hard to extend the
program to emulate calendar functionality.

The String field in Deadline and Event were changed to LocalDateTime
objects, and the string containing the date and time is parsed in the
constructor of the Deadline and Event objects before they are stored.

This allows the Deadline and Event objects to store LocalDateTime
objects, which allows more functionality to be added by using the
LocalDateTime API. By parsing the LocalDateTime in the constructors,
we can prevent code duplication as Storage calls the constructor when
loading Tasks from file. This is achieved by storing the LocalDateTime
in the same string format as it was parsed from.

iamchenjiajun committed:
Add exceptions for date time and handle them

DateTimeFormatException is a subclass of DukeException and is thrown
when the user enters a wrong format for the LocalDateTime field during
parsing. This exception is handled in the main Duke class.

iamchenjiajun committed:
Update whitespace between methods

iamchenjiajun committed:
Add DatedTask interface for tasks with datetime

Filtering is needed to check if a Task is a Deadline or Event to extract
LocalDateTime objects from them. This makes the code hard to extend when
adding new functionality to dated Tasks, or when checking if a Task has
a LocalDateTime field.

An interface called DatedTask has been added which is implemented by
Tasks with a LocalDateTime field.

This enables easier checking of Tasks with LocalDateTime fields and
methods can be added to the interface to extract the field from objects
that implement this interface.

iamchenjiajun committed:
Add date command to filter tasks on a given date

DateCommand is the command executed when user is searching for a date.
DateCommand calls TaskManager.printTasksOnDate, which filters dates
from DatedTasks and prints the tasks.

iamchenjiajun committed:
Update test input/output for new commands

iamchenjiajun committed:
Add find command to search for tasks

iamchenjiajun committed:
Add error string for no find argument

iamchenjiajun committed:
Update find command to be case insensitive

iamchenjiajun committed:
Update test input/output for find command

iamchenjiajun committed:
Add JavaDoc for Storage class.

iamchenjiajun committed:
Add JavaDoc for Parser class.

iamchenjiajun committed:
Add JavaDoc for TaskManager class.

iamchenjiajun committed:
Add JavaDoc for Ui class.

iamchenjiajun committed:
Add JavaDoc for Task class.

iamchenjiajun committed:
Add JavaDoc for DukeException and its subclasses.

iamchenjiajun committed:
Add JavaDoc for Command and its subclasses.

iamchenjiajun committed:
Add JavaDoc for Duke class.

iamchenjiajun committed:
Add JavaDoc for Task subclasses.

iamchenjiajun committed:
Update JavaDoc for Task and its subclasses.

iamchenjiajun committed:
Merge pull request #1 from iamchenjiajun/branch-Level-8

branch-Level-8

iamchenjiajun committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/ui/Ui.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

Retain all changes in Command.java
Retain all changes and modify Ui.java
Retain all changes in input.txt
Retain all changes and modify EXPECTED.TXT

iamchenjiajun committed:
Update test cases

iamchenjiajun committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/command/AddEventCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/taskmanager/TaskManager.java

Retain all changes in Parser.java
Retain all changes in TaskManager.java
Retain all changes and modify Storage.java
Retain all changes and modify Command.java
Retain all changes and modify AddEventCommand.java
Retain all changes and modify Deadline.java
Retain all changes and modify Event.java
Retain all changes and modify Task.java

iamchenjiajun committed:
Remove unused import.

iamchenjiajun committed:
Update JavaDoc for new merged classes.

iamchenjiajun committed:
Merge pull request #2 from iamchenjiajun/branch-Level-9

branch-Level-9

iamchenjiajun committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/taskmanager/TaskManager.java

Retain all changes and modify Parser.java
Retain all changes and modify TaskManager.java

iamchenjiajun committed:
Update JavaDoc for new merged methods

iamchenjiajun committed:
Fix typo in JavaDoc in Parser

iamchenjiajun committed:
Update JavaDoc in DatedTask interface

iamchenjiajun committed:
Add JavaDoc in FindCommand class

iamchenjiajun committed:
Merge pull request #3 from iamchenjiajun/branch-A-JavaDoc

branch-A-JavaDoc

iamchenjiajun committed:
Update User Guide to include features

iamchenjiajun committed:
Set theme jekyll-theme-midnight

iamchenjiajun committed:
Set theme jekyll-theme-cayman

iamchenjiajun committed:
Update GH pages config to use GFMD

iamchenjiajun committed:
Update feature description for README.md

iamchenjiajun committed:
Update GH pages config to use jemoji

iamchenjiajun committed:
Add quick command reference to user guide

iamchenjiajun committed:
Add notes to user guide

iamchenjiajun committed:
Update user guide with more details

iamchenjiajun committed:
Add saving after certain commands

iamchenjiajun committed:
Update user guide with launch information
